src/phillips_dessouky.rs

```diff
@@ -1,80 +1,308 @@
+            // Rounding small negative values to 0.0 avoids pathfinding::edmonds_karp
+            // from entering unreachable code. Has no impact on correctness in test runs.
+            let residual_uv_edge = residual_graph.get_edge_mut(*u, *v);
+            let mut uv_capacity = residual_uv_edge.get_ub() - residual_uv_edge.get_flow();
+            if uv_capacity.into_inner().abs() < self.fp_error {
+                uv_capacity = OrderedFloat(0.0);
+            }
+            residual_uv_edge.set_capacity(uv_capacity);
+
+            let mut vu_capacity = residual_uv_edge.get_flow() - residual_uv_edge.get_lb();
+            if vu_capacity.into_inner().abs() < self.fp_error {
+                vu_capacity = OrderedFloat(0.0);
+            }
+
+            match self.dag.has_edge(*v, *u) {
+                true => residual_graph.get_edge_mut(*v, *u).set_capacity(vu_capacity),
+                false => residual_graph.add_edge(*v, *u, LowtimeEdge::new(
+                    vu_capacity,
+                    OrderedFloat(0.0), // flow
+                    OrderedFloat(0.0), // ub
+                    OrderedFloat(0.0), // lb
+                )),
+            }
+        }
+
+        // Run max flow on the new residual graph.
+        let (flows, _max_flow, _min_cut): (
+            Vec<((u32, u32), OrderedFloat<f64>)>,
+            OrderedFloat<f64>,
+            Vec<((u32, u32), OrderedFloat<f64>)>,
+        ) = residual_graph.max_flow();
+
+        // Convert flows to dict for faster lookup
+        let flow_dict = flows.iter().fold(
+            HashMap::new(),
+            |mut acc: HashMap<u32, HashMap<u32, OrderedFloat<f64>>>, ((from, to), flow)| {
+                acc.entry(*from)
+                    .or_insert_with(HashMap::new)
+                    .insert(*to, *flow);
+                acc
+            }
+        ); 
+
+        // Add additional flow we get to the original graph
+        for (u, v, edge) in self.dag.edges_mut() {
+            edge.incr_flow(*flow_dict.get(u)
+                .and_then(|inner| inner.get(v))
+                .unwrap_or(&OrderedFloat(0.0)));
+            edge.decr_flow(*flow_dict.get(v)
+                .and_then(|inner| inner.get(u))
+                .unwrap_or(&OrderedFloat(0.0)));
+        }
+
+        // Construct the new residual graph.
+        let mut new_residual = self.dag.clone();
+        for (u, v, edge) in self.dag.edges() {
+            new_residual.get_edge_mut(*u, *v).set_flow(edge.get_ub() - edge.get_flow());
+            new_residual.add_edge(*v, *u, LowtimeEdge::new(
+                OrderedFloat(0.0), // capacity
+                edge.get_flow() - edge.get_lb(), // flow
+                OrderedFloat(0.0), // ub
+                OrderedFloat(0.0), // lb
+            ));
+        }
 
-        let fractional_seconds = subsec_nanos as f64 / 1_000_000_000.0;
-        let total_seconds = seconds as f64 + fractional_seconds;
+        if log::log_enabled!(Level::Debug) {
```

Comment:
Can I control the log level by setting the `RUST_LOG` environment variable?

Reply:
[Here}(https://docs.rs/pyo3-log/latest/pyo3_log/#mapping) is some info on `pyo3-log`, but the TLDR is that `pyo3-log` figures out the Python-side logger's configs and maps the Rust-side levels to Python-side levels.

> Can I control the log level by setting the RUST_LOG environment variable?
No, the log level is implicitly set to whatever the Python-side logger is configured for. In the profiling script, I can confirm that changing `level` to `logging.DEBUG` in:
```py
    logging.basicConfig(
        format="%(asctime)s %(levelname)-8s [%(filename)s:%(lineno)d] %(message)s",
        level=logging.DEBUG,
        handlers=[logging.FileHandler(log_path, mode="a"), logging.StreamHandler()],
    )
```
surfaces the Rust-side logs.

On an unrelated note, setting the logging level to DEBUG gave me a bunch of one-time debug logs from the `draw` function. It's a bunch of logs from `font_manager.py:1428` complaining about a bunch (a few hundred from eyeballing) of different fonts. This doesn't seem to affect the drawn images, but wanted to check if this is expected?

Reply:
Usually libraries will do a lot of DEBUG level logging. It is possible to suppress logs from particular packages entirely, so that's what you'd do if you don't want those matplotlib messages. It was something like `getLogger("matplotlib").set_level(logging.WARNING)`. Then every logger whose name begins with `"matplotlib"` will be affected.